src/util/FitsComparator.cpp

```diff
@@ -0,0 +1,208 @@
+//
+// Created by Samuel Cappon on 3/5/20.
+//
+
+#include <util/FitsComparator.h>
+#include <fitsio.h>
+#include <core/Exceptions.h>
+#include <iostream>
+#include <memory>
+#include <vector>
+#include <map>
+#include <algorithm>
+#include <iterator>
+#include <core/SimpleLogger.h>
+
+using namespace std;
+
+FitsComparator::FitsComparator(string fitsfile1, string fitsfile2) {
+    int status = 0;
+    int hdunum;
+
+        fits_open_file(&_fptr, fitsfile1.c_str(), READONLY, &status);
+        fits_open_file(&_fptr2, fitsfile2.c_str(), READONLY, &status);
+        if(status){
+            string err = "Failed to open files " + fitsfile1 + " and " + fitsfile2 + ", check paths and extensions.";
+            LOG_CRT << err;
+            throw L1::CfitsioError(err);
+        }
+        fits_get_num_hdus(_fptr, &hdunum, &status);
+        if(hdunum != 17){
+            string warning = "File " + fitsfile1 + " does not have 17 hdu's. Total hdu = " + to_string(hdunum);
+            LOG_CRT << warning;
+        }
+        fits_get_num_hdus(_fptr2, &hdunum, &status);
+        if(hdunum != 17){
+            string warning = "File " + fitsfile2 + " does not have 17 hdu's. Total hdu = " + to_string(hdunum);
+            LOG_CRT << warning;
+        }
+}
+
+FitsComparator::~FitsComparator() {
+    int status = 0;
+    fits_close_file(_fptr, &status);
+    fits_close_file(_fptr2, &status);
+}
+
+vector<string> FitsComparator::get_segments(fitsfile *fptr) {
+    int status = 0, i;
+    char segment_value[FLEN_VALUE], segment_comment[FLEN_COMMENT];
+    vector<string> segments;
+
+    for (i = 2; i <= 17; i++) {
+        fits_movabs_hdu(fptr, i, NULL, &status);
+        fits_read_key(fptr, TSTRING, "EXTNAME", segment_value, segment_comment, &status);
+        if(status){
+            string err = "Cannot read EXTNAME key from hdu " + to_string(i);
+            LOG_CRT << err;
+            throw L1::CfitsioError(err);
+        }
+        string segment = string(segment_value);
+        segments.push_back(segment);
+    }
+    return segments;
+}
+
+multimap<string, string> FitsComparator::get_header_values(fitsfile *fptr){
+    int status = 0, i, nkeys;
+    char keyname[FLEN_KEYWORD], value[FLEN_VALUE], comment[FLEN_COMMENT];
+
+    fits_get_hdrspace(fptr, &nkeys, NULL, &status);
+
+    multimap<string, string> header_values;
+
+    for(i = 1; i <= nkeys; i++){
+        fits_read_keyn(fptr, i, keyname, value, comment, &status);
+        header_values.insert(pair<string, string>(keyname, value));
+    }
+    return header_values;
+}
+
+
+vector<vector<int32_t>> FitsComparator::get_pixels(fitsfile *fptr){
+    int bitpix, naxis, i, status = 0;
+    long naxes[2];
+    long fpixel[2];
+
+    fits_get_img_param(fptr, 2, &bitpix, &naxis, naxes, &status);
+    if(status){
```

Comment:
This block should probably be moved to end of the function because when status > 0, fits function calls stop executing the commands.
1. you should check status for every fits function call
2. since it sets once and stops executing after error occurred, this check should be end of function.